### PR TITLE
DEPR: Series/DataFrame.append (#35407)

### LINKED ISSUE #22957: How should `DataFrame.append` behave related to indexes types?
I'm currently working on a refactoring the code `DataFrame.append` (#22915). One question that came up was, what should be the behavior when appending DataFrames with different index types.

To simplify this, let's put the focus on the index itself, and assume that it is valid for both rows and columns (IMHO consistency between the two might be good).

### Code Sample

```python
>>> df1 = pd.DataFrame([[1, 2, 3]], columns=index1)
>>> df2 = pd.DataFrame([[4, 5, 6]], columns=index2)
>>> # what should happen in the next line when index1 and
>>> # index2 are of different types?
>>> result = df1.append(df2)
```

#### Index 0 (rows)

All types of indexes work together, except for:

- `CategoricalIndex + {another}`

When the types don't match, the result is cast to `object`.

There's a also a bug that happend when appending MultiIndex + DatetimeIndex. I will verify it more in detail later and raise an issue.

#### Index 1 (columns)

All types of indexes work together, except for:

- `IntervalIndex + {another}`
- `MultiIndex + {another}`
- `{another} + PeriodIndex`
- `{another} + IntervalIndex`
-  `{another} + MultiIndex`
- `DatetimeIndex + {another}` (works for `df.append(series)`)
- `TimedeltaIndex + {numeric}` (works for `df.append(series)`)
- `PeriodIndex + {another}` (works for `df.append(series)`)

The kinds of exceptions raised here are the most varied and sometimes very cryptic.

For example:

```python
>>> df1 = pd.DataFrame([[1]], columns=['A'])
>>> df2 = pd.DataFrame([[2]], columns=pd.PeriodIndex([2000], freq='A'))
>>> df1.append(df2)
Traceback (most recent call last):
  File "<ipython-input-15-8ab0723181fb>", line 1, in <module>
    df1.append(df2)
  File "/home/andre/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/frame.py", line 6211, in append
    sort=sort)
  File "/home/andre/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/reshape/concat.py", line 226, in concat
    return op.get_result()
  File "/home/andre/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/reshape/concat.py", line 423, in get_result
    copy=self.copy)
  File "/home/andre/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/internals.py", line 5416, in concatenate_block_managers
    elif is_uniform_join_units(join_units):
  File "/home/andre/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/internals.py", line 5440, in is_uniform_join_units
    all(not ju.is_na or ju.block.is_extension for ju in join_units) and
  File "/home/andre/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/internals.py", line 5440, in <genexpr>
    all(not ju.is_na or ju.block.is_extension for ju in join_units) and
AttributeError: 'NoneType' object has no attribute 'is_extension'
```

### Suggestion

Since index 0 already allows almost all types of index the be concatenated, we don't want to restrict this behavior. A suggestion I give is to allow any types of indexes to be merged together (and possibly raise a RuntimeWarning when a cast is necessary).

Foe example, joining two indexes of the same type should produce the same type (as the current behavior):

```python
>>> df1 = pd.DataFrame([[1]], columns=pd.Int64Index([1]))
>>> df2 = pd.DataFrame([[2]], columns=pd.Int64Index([2]))
>>> result = df1.append(df2)
>>> result
     1    2
0  1.0  NaN
0  NaN  2.0
>>> result.columns
Int64Index([1, 2], dtype='int64')
```

Joining two indexes of different types usually upcasts to `object` (only when necessary)

```python
>>> df1 = pd.DataFrame([[1]], columns=pd.Int64Index([1]))
>>> df2 = pd.DataFrame([[2]], columns=pd.interval_range(0, 1))
>>> result = df1.append(df2)
>>> result
     1  (0, 1]
0  1.0     NaN
1  NaN     2.0
>>> result.columns
Index([1, (0, 1]], dtype='object')
```

### Empty indexes

I believe that empty indexes (e.g. those created in a `DataFrame` with no rows or columns) should be ignored when calculating the final dtype of an index merge. This seems like it is already the current behavior:

```python
>>> df1 = pd.DataFrame()
>>> df2 = pd.DataFrame(index=[1])
>>> df1.index
Index([], dtype='object')
>>> df2.index
Int64Index([1], dtype='int64')
>>> df1.append(df2).index
Int64Index([1], dtype='int64')
```

However, when an empty index has a dtype different from `object`, we may want to preserve it (as it may have been created explicitly by the user).

```python
>>> # current behavior
>>> df1 = pd.DataFrame(index=pd.Float64Index([]))
>>> df2 = pd.DataFrame(index=[1])
>>> df1.append(df2).index
Int64Index([1], dtype='int64')
>>>
>>> # suggested behavior
>>> df1.append(df2).index
Float64Index([1], dtype='int64')
```

---

Sorry if this was too long, this is being my first contribution to open source and I still didn't get the hang of how things work. Any suggestions, whether related to the issue or meta, are welcome!

### LINKED ISSUE #35407: Deprecate Series / DataFrame.append
I think that we should deprecate `Series.append` and `DataFrame.append`. They're making an analogy to `list.append`, but it's a poor analogy since the behavior isn't (and can't be) in place. The data for the index and values needs to be copied to create the result.

These are also apparently popular methods. DataFrame.append is around the 10th most visited page in our API docs.

Unless I'm mistaken, users are always better off building up a list of values and passing them to the constructor, or building up a list of NDFrames followed by a single `concat`.

### PR DESCRIPTION
**Pull request by neinkeinkaffee**
- [x] closes #35407 (adds deprecation warning)
- [x] closes #22957 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -438,6 +438,61 @@ when given numeric data, but in the future, a :class:`NumericIndex` will be retu
...
+    0  1  2
+    1  3  4
+    0  5  6
+    1  7  8
+
+*Recommended syntax*
+
+.. code-block:: ipython
+
+    In [1]: pd.core.reshape.concat.concat([pd.Series([1, 2]), pd.Series([3, 4])])
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:477]: use the idiomatic pd.concat

File: pandas/core/frame.py [hunk 2]
```diff
@@ -9114,6 +9114,14 @@ def append(
         3  3
         4  4
         """
+        warnings.warn(
+            "The frame.append method is deprecated "
+            "and will be removed from pandas in a future version. "
+            "Use pandas.core.reshape.concat instead.",
+            FutureWarning,
+            stacklevel=2,
```
Comment (jreback) [inline pandas/core/frame.py:9122]: use find_stack_level

File: pandas/core/series.py [hunk 3]
```diff
@@ -2888,6 +2888,14 @@ def append(
         ...
         ValueError: Indexes have overlapping values: [0, 1, 2]
         """
+        warnings.warn(
+            "The frame.append method is deprecated "
+            "and will be removed from pandas in a future version. "
+            "Use pandas.core.reshape.concat instead.",
+            FutureWarning,
+            stacklevel=2,
```
Comment (jreback) [inline pandas/core/series.py:2896]: same

File: pandas/tests/frame/methods/test_append.py [hunk 4]
```diff
@@ -21,19 +21,22 @@ def test_append_multiindex(self, multiindex_dataframe_random_data, frame_or_seri
         a = obj[:5]
         b = obj[5:]
 
-        result = a.append(b)
+        with tm.assert_produces_warning(FutureWarning):
```
Comment (jreback) [inline pandas/tests/frame/methods/test_append.py:24]: would be ok just to filter warnings for this entire class

**Review by jreback — CHANGES_REQUESTED** [review]
ok to filter warnings and have a single test to assert the deprecation warning

**Comment by neinkeinkaffee** [comment]
Thanks for the suggestions. I've tried to address them. Could you re-review and let me know your feedback? Cheers!

File: pandas/core/series.py [hunk 5]
```diff
@@ -2888,6 +2888,14 @@ def append(
         ...
         ValueError: Indexes have overlapping values: [0, 1, 2]
         """
+        warnings.warn(
+            "The frame.append method is deprecated "
```
Comment (phofl) [inline pandas/core/series.py:2892]: Series here

File: pandas/tests/frame/methods/test_append.py [hunk 6]
```diff
@@ -270,3 +270,9 @@ def test_multiindex_column_append_multiple(self):
                 {"col1": [1, 2, 3] * (i + 1), "col2": [11, 12, 13] * (i + 1)}
             )
             tm.assert_frame_equal(result, expected)
+
+    def test_append_raises_future_warning(self):
```
Comment (phofl) [inline pandas/tests/frame/methods/test_append.py:274]: add issue number

File: pandas/tests/series/methods/test_append.py [hunk 7]
```diff
@@ -82,7 +83,12 @@ def test_append_dataframe_raises(self):
         with pytest.raises(TypeError, match=msg):
             df.A.append([df])
 
+    def test_append_raises_future_warning(self):
+        with tm.assert_produces_warning(FutureWarning):
```
Comment (phofl) [inline pandas/tests/series/methods/test_append.py:87]: issue number

File: doc/source/whatsnew/v1.4.0.rst [hunk 8]
```diff
@@ -438,6 +438,61 @@ when given numeric data, but in the future, a :class:`NumericIndex` will be retu
...
+:meth:`DataFrame.append` and :meth:`Series.append` have been deprecated and will be removed in Pandas 2.0.
+Use :func:`pandas.concat` instead (:issue:`35407`).
+
+*Deprecated syntax*
+
+.. code-block:: ipython
+
+    In [1]: pd.Series([1, 2]).append(pd.Series([3, 4])
+    Out [1]:
+    <stdin>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
```
Comment (phofl) [inline doc/source/whatsnew/v1.4.0.rst:455]: series.append

**Review by phofl — CHANGES_REQUESTED** [review]

File: pandas/tests/series/methods/test_append.py (hunk 7 above)
Comment (neinkeinkaffee) [reply pandas/tests/series/methods/test_append.py:87]: Thanks! Where does the issue number go? Is there another example for this in the existing tests?

Comment (MarcoGorelli) [reply pandas/tests/series/methods/test_append.py:87]: Just below the definition, e.g.
```
def test_append_raises_future_warning(self):
    # GH#44539
```

**Review by phofl — CHANGES_REQUESTED** [review]
Web and docs is failing because of the warnings

**Comment by neinkeinkaffee** [comment]
Question on the failing web and docs: looking at the failing doc build, some of the cases are due to direct calls to append, which I can change into concat. But others are indirect, e.g. `doc/source/info_tutorials/getting_started/02_read_write.rst` calls titanic.info() on a DataFrame, which calls frame.memory_usage which calls series.append. In such cases, it's probably preferred to change the series.append into a concat in frame.memory_usage? Or should I just add an `:okwarning:` to the ipython codeblock?

**Comment by jreback** [comment]
no all usages should be changes in tests and docs except those we r explicitly doing

**Comment by neinkeinkaffee** [comment]
@jreback Just to make sure I don't misunderstand: given that frame.memory_usage is neither in doc nor test, I shouldn't change the call to series.append to concat there, but instead add a :okwarning: to the ipython codeblock? And likewise not change any other instances were append is called in the pandas code that's neither documentation nor test?

**Comment by jreback** [comment]
no u do not add okwarning anywhere 

u need to fix this for example in the code not to output a warning (as this is tested in many tests)

**Comment by jreback** [comment]
when you push pls merge master

**Comment by neinkeinkaffee** [comment]
Yes, will do!

File: doc/source/whatsnew/v1.4.0.rst [hunk 9]
```diff
@@ -490,6 +490,61 @@ when given numeric data, but in the future, a :class:`NumericIndex` will be retu
...
+    <stdin>:1: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
+       A  B
+    0  1  2
+    1  3  4
+    0  5  6
+    1  7  8
+
+*Recommended syntax*
+
+.. code-block:: ipython
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:527]: dont' use a code-block here, use an ipython-block (so it runs the code)

File: pandas/core/indexing.py [hunk 10]
```diff
@@ -1993,7 +1993,14 @@ def _setitem_with_indexer_missing(self, indexer, value):
                     df = df.infer_objects()
                 self.obj._mgr = df._mgr
             else:
-                self.obj._mgr = self.obj.append(value)._mgr
+                from pandas.core.reshape.concat import concat
+
+                df = value.to_frame().T
```
Comment (jreback) [inline pandas/core/indexing.py:1998]: hmm this not great. cc @jbrockmendel can you suggest anything better here

File: pandas/tests/base/test_value_counts.py [hunk 11]
```diff
@@ -38,6 +38,7 @@ def test_value_counts(index_or_series_obj):
     tm.assert_series_equal(result, expected)
 
 
+@pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/base/test_value_counts.py:41]: why is this showing a warning? generally *don't* add filterwarnings in anything but the actual method testing itself (e.g. append), these can be suppressing a warning that is thrown internally which we want to fix

File: pandas/tests/frame/methods/test_append.py [hunk 12]
```diff
@@ -12,6 +12,7 @@
 import pandas._testing as tm
 
 
+@pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/frame/methods/test_append.py:15]: this is ok!

File: pandas/tests/frame/methods/test_drop_duplicates.py [hunk 13]
```diff
@@ -21,6 +21,7 @@ def test_drop_duplicates_with_misspelled_column_name(subset):
         df.drop_duplicates(subset)
 
 
+@pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/frame/methods/test_drop_duplicates.py:24]: not here

File: pandas/tests/generic/test_duplicate_labels.py [hunk 14]
```diff
@@ -294,6 +294,7 @@ def test_setting_allows_duplicate_labels_raises(self, data):
 
         assert data.flags.allows_duplicate_labels is True
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/generic/test_duplicate_labels.py:297]: not here!

File: pandas/tests/generic/test_finalize.py [hunk 15]
```diff
@@ -511,6 +511,7 @@ def ndframe_method(request):
     return request.param
 
 
+@pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/generic/test_finalize.py:514]: not here! (if append is in this list then ok)

File: pandas/tests/indexes/period/test_indexing.py [hunk 16]
```diff
@@ -115,6 +115,7 @@ def test_getitem_index(self):
         )
         tm.assert_index_equal(result, exp)
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/indexes/period/test_indexing.py:118]: not here!

**Review by jreback — CHANGES_REQUESTED** [review]
you don't simply want to suppress the warnings anywhere, you need to actually fix code that is causing the warnings.

File: pandas/core/frame.py [hunk 17]
```diff
@@ -3265,9 +3265,16 @@ def memory_usage(self, index: bool = True, deep: bool = False) -> Series:
...
         )
         if index:
-            result = self._constructor_sliced(
-                self.index.memory_usage(deep=deep), index=["Index"]
-            ).append(result)
+            from pandas.core.reshape.concat import concat
+
+            result = concat(
+                [
+                    self._constructor_sliced(
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3272]: nitpick: can you do `whatever = self._constructor_sliced(...)` in a separate line

File: pandas/core/frame.py [hunk 18]
```diff
@@ -9709,7 +9724,11 @@ def c(x):
             idx_diff = result_index.difference(correl.index)
 
             if len(idx_diff) > 0:
-                correl = correl.append(Series([np.nan] * len(idx_diff), index=idx_diff))
+                from pandas.core.reshape.concat import concat
+
+                correl = concat(
+                    [correl, Series([np.nan] * len(idx_diff), index=idx_diff)]
```
Comment (jbrockmendel) [inline pandas/core/frame.py:9730]: can you define the extra row on its own line (ideally with an informative name)

File: pandas/core/indexing.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/core/indexing.py:1998]: this looks very nearly cut/paste from DataFrame.append.  the only difference is that we actually have _tests_ for DataFrame.append.  If this code has to exist somewhere, then we might as well just implement DataFrame._append so we can adapt the existing tests (and im guessing the scope of DataFrame._append would be narrower so once the deprecation is enforced we could rip out a bunch of it)

**Comment by MarcoGorelli** [comment]
I think @jbrockmendel 's suggestion

> this looks very nearly cut/paste from DataFrame.append. the only difference is that we actually have tests for DataFrame.append. If this code has to exist somewhere, then we might as well just implement DataFrame._append so we can adapt the existing tests (and im guessing the scope of DataFrame._append would be narrower so once the deprecation is enforced we could rip out a bunch of it)

makes a lot of sense.

So, this would mean:
- rename `append` to `_append`
- inside `append`, just raise a future warning and then call `_append`
- when `append` is used in internal code, replace it with `_append`
- when `append` is used in user-facing code (e.g. docs), replace it with `concat`
?

@neinkeinkaffee apologies, this has ended up being a lot more work than I'd anticipated when I suggested the issue. However, hopefully you're finding it to be worthwhile!

**Comment by jreback** [comment]
+1 on @jbrockmendel suggestion

**Comment by neinkeinkaffee** [comment]
I find it absolutely worthwhile, thanks @MarcoGorelli. 

Thanks everyone for your comments and patience, really appreciated!

File: pandas/core/frame.py [hunk 19]
```diff
@@ -3267,9 +3267,17 @@ def memory_usage(self, index: bool = True, deep: bool = False) -> Series:
...
+            index_memory_usage = self._constructor_sliced(
                 self.index.memory_usage(deep=deep), index=["Index"]
-            ).append(result)
+            )
+            result = concat(
+                [
+                    index_memory_usage,
+                    result,
+                ]
+            )
```
Comment (MarcoGorelli) [inline pandas/core/frame.py:3280]: really minor, but can this fit on a single line?

File: pandas/tests/generic/test_duplicate_labels.py (hunk 14 above)
Comment (jreback) [reply pandas/tests/generic/test_duplicate_labels.py:297]: remove this

File: pandas/tests/indexing/test_partial.py [hunk 20]
```diff
@@ -361,7 +361,7 @@ def test_partial_setting_mixed_dtype(self):
 
         s = df.loc[1].copy()
         s.name = 2
-        expected = df.append(s)
+        expected = df._append(s)
```
Comment (jreback) [inline pandas/tests/indexing/test_partial.py:364]: can you try not to use _append in tests and pref rewriting as concat

File: pandas/tests/reshape/concat/test_categorical.py [hunk 21]
```diff
@@ -184,6 +185,7 @@ def test_concat_categorical_unchanged(self):
         )
         tm.assert_equal(result, expected)
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/reshape/concat/test_categorical.py:188]: remove

File: pandas/tests/reshape/concat/test_concat.py [hunk 22]
```diff
@@ -219,6 +219,7 @@ def test_concat_keys_levels_no_overlap(self):
         with pytest.raises(ValueError, match=msg):
             concat([df, df2], keys=["one", "two"], levels=[["foo", "bar", "baz"]])
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/reshape/concat/test_concat.py:222]: remove

File: pandas/tests/reshape/concat/test_index.py [hunk 23]
```diff
@@ -141,6 +141,7 @@ def test_default_index(self):
         exp = DataFrame([[1, 5, 3, 7], [2, 6, 4, 8]])
         tm.assert_frame_equal(res, exp, check_index_type=True, check_column_type=True)
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/reshape/concat/test_index.py:144]: remove

File: pandas/tests/reshape/merge/test_merge.py [hunk 24]
```diff
@@ -682,6 +682,7 @@ def _constructor(self):
 
         assert isinstance(result, NotADataFrame)
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py:685]: remove

**Review by jreback — CHANGES_REQUESTED** [review]
remove almost all of the generic filtering, the *only* ones where it should remain is the explict test_append.py

File: pandas/tests/generic/test_duplicate_labels.py [hunk 25]
```diff
@@ -294,15 +294,11 @@ def test_setting_allows_duplicate_labels_raises(self, data):
...
-    @pytest.mark.parametrize(
-        "func", [operator.methodcaller("append", pd.Series(0, index=["a", "b"]))]
-    )
-    def test_series_raises(self, func):
+    def test_series_raises(self):
         s = pd.Series([0, 1], index=["a", "b"]).set_flags(allows_duplicate_labels=False)
         msg = "Index has duplicates."
         with pytest.raises(pd.errors.DuplicateLabelError, match=msg):
-            func(s)
+            pd.concat([pd.Series(0, index=["a", "b"]), s])
```
Comment (jbrockmendel) [inline pandas/tests/generic/test_duplicate_labels.py:301]: can you construct the series outside of the pytest.raises; makes clearer exactly what part of this line is raising

File: pandas/tests/indexing/test_partial.py [hunk 26]
```diff
@@ -538,7 +538,8 @@ def test_partial_set_invalid(self):
         # allow object conversion here
         df = orig.copy()
         df.loc["a", :] = df.iloc[0]
-        exp = orig._append(Series(df.iloc[0], name="a"))
+        s = Series(df.iloc[0], name="a")
```
Comment (jbrockmendel) [inline pandas/tests/indexing/test_partial.py:541]: "s" -> "ser" pls

File: pandas/tests/reshape/concat/test_index.py [hunk 27]
```diff
@@ -179,16 +178,12 @@ def test_dups_index(self):
         tm.assert_frame_equal(result.iloc[10:], df)
 
         # append
-        result = df.iloc[0:8, :].append(df.iloc[8:])
+        result = concat([df.iloc[0:8, :], df.iloc[8:]])
```
Comment (jbrockmendel) [inline pandas/tests/reshape/concat/test_index.py:181]: are these redundant with the tests above?  this _may_ be better thought of as a targeted test for _append.

File: pandas/core/frame.py [hunk 28]
```diff
@@ -3267,9 +3267,12 @@ def memory_usage(self, index: bool = True, deep: bool = False) -> Series:
...
         )
         if index:
-            result = self._constructor_sliced(
+            from pandas.core.reshape.concat import concat
+
+            index_memory_usage = self._constructor_sliced(
                 self.index.memory_usage(deep=deep), index=["Index"]
-            ).append(result)
+            )
+            result = concat([index_memory_usage, result])
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3275]: i'd rather keep using _append internally; should be more performant and avoids circular dependency

File: pandas/tests/reshape/concat/test_append.py [hunk 29]
```diff
@@ -109,7 +110,8 @@ def test_append_sorts(self, sort):
         df1 = DataFrame({"a": [1, 2], "b": [1, 2]}, columns=["b", "a"])
         df2 = DataFrame({"a": [1, 2], "c": [3, 4]}, index=[2, 3])
 
-        with tm.assert_produces_warning(None):
+        # GH#35407
+        with tm.assert_produces_warning(FutureWarning):
             result = df1.append(df2, sort=sort)
```
Comment (jbrockmendel) [inline pandas/tests/reshape/concat/test_append.py:115]: if this file is about testing 'append', then these should just be changed to test '_append'

**Comment by jreback** [comment]
@neinkeinkaffee can you merge master

File: doc/source/user_guide/merging.rst [hunk 30]
```diff
@@ -473,14 +407,20 @@ like GroupBy where the order of a categorical variable is meaningful.
 Appending rows to a DataFrame
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-While not especially efficient (since a new object must be created), you can
-append a single row to a ``DataFrame`` by passing a ``Series`` or dict to
-``append``, which returns a new ``DataFrame`` as above.
+You can append a single row in form of a series to a ``DataFrame`` in-place
```
Comment (jreback) [inline doc/source/user_guide/merging.rst:410]: don't show L416, just make the alternative the thing

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment, ping on green

**Comment by jbrockmendel** [comment]
closes #22957?

**Comment by neinkeinkaffee** [comment]
@jreback I've changed the docs and removed the part about adding rows in-place. However, the Database / Linux_py38_IO pipeline currently fails with `pytest: command not found`, while the Azure pipelines (pandas-dev.pandas Windows py38_np18_2 and Windows py39_2), fail because of a DeprecationWarning that gets raised in `pandas/tests/io/parser/test_header.py::test_no_header_prefix[pyarrow]`, which doesn't appear to be related to the changes I've made. Is this something known or should I dig into it?

**Comment by jreback** [comment]
@neinkeinkaffee if you can merge master again; just want to get a clean run here. we do have some occasional test failures on CI (though some recent PRs do address some of these)

**Review by jreback — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]
LGTM. Great job @neinkeinkaffee

**Comment by jreback** [comment]
thanks @neinkeinkaffee yep very nice!

**Pull request merged**